File: src/lzx.py
"""lzxd_helper discovery and invocation.

The helper is a small C program that links libmspack's internal lzxd_*
API and decodes the raw Turn 10 LZX framing used by FM4's bin.zip (same
Xbox-360 LZX flavour as FH1). See ``src/lzxd_helper.c`` and ``release.sh``
for how it is built.

Search order:
  1. $FORZAMANIA_LZXD_HELPER (explicit override)
  2. sys._MEIPASS/lzxd_helper            (PyInstaller one-file bundle)
  3. <exe_dir>/lzxd_helper               (shipped next to the executable)
  4. <repo>/src/lzxd_helper              (dev: built in place by release.sh)

Decompression goes through a single long-running daemon process: one
``fork``/``exec`` at first use, then a binary stdin/stdout protocol per
request. Saves ~3 ms of fork+exec overhead per LZX entry — the main
cost driver when extracting tens of thousands of bin.zip entries.
"""
from __future__ import annotations

import atexit
import os
import struct
import subprocess
import sys
import threading
from functools import lru_cache
from pathlib import Path


LZX_WINDOW_BITS = 17
LZX_RESET_INTERVAL = 0
LZX_CHUNK_USIZE = 32768
LZX_TRAILER = 5


class DecompressionError(RuntimeError):
    pass


@lru_cache(maxsize=1)
def helper_path() -> Path:
    override = os.environ.get("FORZAMANIA_LZXD_HELPER")
    if override:
        p = Path(override)
        if not p.exists():
            raise FileNotFoundError(f"FORZAMANIA_LZXD_HELPER points at {p} which does not exist")
        return p

    meipass = getattr(sys, "_MEIPASS", None)
    if meipass:
        p = Path(meipass) / "lzxd_helper"
        if p.exists():
            return p

    exe_dir = Path(sys.executable).resolve().parent
    here = Path(__file__).resolve()
    src_dir = here.parent                        # .../src (lzx.py lives in src/)
    candidates = [
        exe_dir / "lzxd_helper",
        src_dir / "lzxd_helper",
    ]
    for c in candidates:
        if c.exists():
            return c.resolve()
    raise FileNotFoundError(
        "lzxd_helper not found. Run release.sh --local or set FORZAMANIA_LZXD_HELPER."
    )


def strip_chunk_headers(blob: bytes, uncomp_size: int) -> bytes:
    """Parse Turn 10's multi-chunk LZX framing into a continuous bitstream.

    Single-chunk entries: ``FF [u16 BE uncomp] [u16 BE comp] <stream> [5-byte trailer]``.
    Multi-chunk entries: each non-last chunk is prefixed with ``u16 BE csize``;
    the last chunk uses the FF-prelude + trailer form.
    """
    out = bytearray()
    pos = 0
    remaining = uncomp_size
    while remaining > 0:
        last = remaining <= LZX_CHUNK_USIZE
        if last:
            if pos + 5 > len(blob):
                raise DecompressionError(
                    f"truncated final chunk header at pos={pos} (need 5, have {len(blob)-pos})"
                )
            if blob[pos] != 0xFF:
                raise DecompressionError(
                    f"expected 0xFF at final-chunk pos={pos}, got 0x{blob[pos]:02x}"
                )
            u = int.from_bytes(blob[pos + 1:pos + 3], "big")
            c = int.from_bytes(blob[pos + 3:pos + 5], "big")
            pos += 5
            end = pos + c
            if end + LZX_TRAILER > len(blob):
                raise DecompressionError(
                    f"truncated final chunk body: comp={c} trailer={LZX_TRAILER} "
                    f"pos={pos} len={len(blob)}"
                )
            out.extend(blob[pos:end])
            pos = end + LZX_TRAILER
            step = u
        else:
            if pos + 2 > len(blob):
                raise DecompressionError(f"truncated chunk header at pos={pos}")
            c = int.from_bytes(blob[pos:pos + 2], "big")
            pos += 2
            end = pos + c
            if end > len(blob):
                raise DecompressionError(
                    f"truncated chunk body: csize={c} pos={pos} len={len(blob)}"
                )
            out.extend(blob[pos:end])
            pos = end
            step = LZX_CHUNK_USIZE
        remaining -= step
    if pos != len(blob):
        raise DecompressionError(f"framing drift: ended at pos={pos}, blob len={len(blob)}")
    return bytes(out)


# ---- daemon client -------------------------------------------------------

_daemon_proc: subprocess.Popen | None = None
_daemon_lock = threading.Lock()
_HDR = struct.Struct(">II")


def _spawn_daemon() -> subprocess.Popen:
    helper = helper_path()
    proc = subprocess.Popen(
        [str(helper), "daemon"],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=None,                # let the user see helper diagnostics
        bufsize=0,                  # we manage buffering ourselves
    )
    if proc.stdin is None or proc.stdout is None:
        raise DecompressionError("failed to attach pipes to lzxd_helper daemon")
    atexit.register(_shutdown_daemon)
    return proc


def _shutdown_daemon() -> None:
    global _daemon_proc
    p = _daemon_proc
    _daemon_proc = None
    if p is None:
        return
    try:
        if p.stdin and not p.stdin.closed:
            p.stdin.close()
        p.wait(timeout=5)
    except Exception:
        try:
            p.kill()
        except Exception:
            pass


def _read_exact(fp, n: int) -> bytes:
    """Read exactly ``n`` bytes from a binary pipe; raise on short read."""
    if n == 0:
        return b""
    chunks: list[bytes] = []
    got = 0
    while got < n:
        b = fp.read(n - got)
        if not b:
            raise DecompressionError(
                f"lzxd_helper daemon closed pipe after {got}/{n} bytes"
            )
        chunks.append(b)
        got += len(b)
    return b"".join(chunks) if len(chunks) > 1 else chunks[0]


def decode_lzx(stream: bytes, out_len: int) -> bytes:
    """Decompress one stripped LZX bitstream via the helper daemon.

    The stream must already have its multi-chunk framing removed (use
    ``strip_chunk_headers`` for ``bin.zip`` entries). ``out_len`` is the
    expected uncompressed length and is required by libmspack's lzxd
    decoder.
    """
    global _daemon_proc
    with _daemon_lock:
        if _daemon_proc is None or _daemon_proc.poll() is not None:
            _daemon_proc = _spawn_daemon()
        proc = _daemon_proc

        try:
            proc.stdin.write(_HDR.pack(out_len, len(stream)))
            if stream:
                proc.stdin.write(stream)
            proc.stdin.flush()

            hdr = _read_exact(proc.stdout, _HDR.size)
            status, payload_len = _HDR.unpack(hdr)
            payload = _read_exact(proc.stdout, payload_len)
        except (BrokenPipeError, DecompressionError):
            # Helper crashed or closed. Reap and surface the error.
            try:
                proc.kill()
            except Exception:
                pass
            _daemon_proc = None
            raise DecompressionError("lzxd_helper daemon died mid-request")

        if status != 0:
            raise DecompressionError(
                f"lzxd_helper: {payload.decode('utf-8', errors='replace').strip()}"
            )
        if len(payload) != out_len:
            raise DecompressionError(
                f"lzxd output length {len(payload)} != expected {out_len}"
            )
        return payload

File: app/lzx.nim
## lzxd_helper discovery, Turn 10 LZX chunk-framing, and the decode daemon.
##
## Nim port of `src/lzx.py`. The helper is a small C program linking libmspack's
## internal lzxd_* API; it decodes the Xbox-360 LZX flavour FM4's bin.zip uses.
## Decompression goes through one long-running daemon process (one fork/exec per
## extraction batch, then a binary stdin/stdout protocol per entry) — the same
## amortisation the Python side does over tens of thousands of entries.
##
## Daemon wire protocol (see src/lzxd_helper.c, big-endian u32s):
##   request : u32 out_len, u32 in_len, in_len bytes of stripped LZX stream
##   response: u32 status (0=ok), u32 payload_len, payload_len bytes
##             (decompressed bytes on status 0, else an ASCII error message)

import std/[os, osproc, streams, strutils]

const
  LZX_CHUNK_USIZE = 32768
  LZX_TRAILER     = 5

type
  DecompressionError* = object of CatchableError
  LzxDaemon* = object
    process*: Process
    inp: Stream
    outp: Stream

when defined(windows):
  const helperNames = ["lzxd_helper.exe", "lzxd_helper"]
else:
  const helperNames = ["lzxd_helper"]

proc findHelper*(): string =
  ## Locate the lzxd_helper binary. Search order mirrors src/lzx.py:
  ##   1. $FORZAMANIA_LZXD_HELPER override
  ##   2. <appdir>/tools/lzxd_helper   (release bundle layout)
  ##   3. <appdir>/lzxd_helper         (beside the binary)
  let override = getEnv("FORZAMANIA_LZXD_HELPER")
  if override.len > 0:
    if not fileExists(override):
      raise newException(IOError,
        "FORZAMANIA_LZXD_HELPER points at " & override & " which does not exist")
    return override
  let appDir = getAppDir()
  for d in [appDir / "tools", appDir]:
    for name in helperNames:
      let p = d / name
      if fileExists(p): return p
  raise newException(IOError,
    "lzxd_helper not found in " & (appDir / "tools") & " or " & appDir &
    ". Build it (release.sh --local) or set FORZAMANIA_LZXD_HELPER.")

# ---- chunk framing -------------------------------------------------------

proc stripChunkHeaders*(blob: string; uncompSize: int): string =
  ## Parse Turn 10's multi-chunk LZX framing into a continuous bitstream.
  ## Single-chunk entries: FF [u16 BE uncomp] [u16 BE comp] <stream> [5-byte trailer].
  ## Non-last chunks are prefixed with u16 BE csize; the last uses the FF form.
  var
    res = newStringOfCap(uncompSize)
    pos = 0
    remaining = uncompSize
  template b(i: int): int = blob[i].byte.int
  while remaining > 0:
    let last = remaining <= LZX_CHUNK_USIZE
    var step: int
    if last:
      if pos + 5 > blob.len:
        raise newException(DecompressionError, "truncated final chunk header at pos=" & $pos)
      if b(pos) != 0xFF:
        raise newException(DecompressionError,
          "expected 0xFF at final-chunk pos=" & $pos & ", got 0x" & toHex(b(pos), 2))
      let u = (b(pos+1) shl 8) or b(pos+2)
      let c = (b(pos+3) shl 8) or b(pos+4)
      pos += 5
      let endp = pos + c
      if endp + LZX_TRAILER > blob.len:
        raise newException(DecompressionError, "truncated final chunk body: comp=" & $c)
      res.add(blob[pos ..< endp])
      pos = endp + LZX_TRAILER
      step = u
    else:
      if pos + 2 > blob.len:
        raise newException(DecompressionError, "truncated chunk header at pos=" & $pos)
      let c = (b(pos) shl 8) or b(pos+1)
      pos += 2
      let endp = pos + c
      if endp > blob.len:
        raise newException(DecompressionError, "truncated chunk body: csize=" & $c)
      res.add(blob[pos ..< endp])
      pos = endp
      step = LZX_CHUNK_USIZE
    remaining -= step
  if pos != blob.len:
    raise newException(DecompressionError,
      "framing drift: ended at pos=" & $pos & ", blob len=" & $blob.len)
  return res

# ---- daemon client -------------------------------------------------------

proc writeU32BE(s: Stream; v: uint32) =
  s.write(char((v shr 24) and 0xFF))
  s.write(char((v shr 16) and 0xFF))
  s.write(char((v shr 8) and 0xFF))
  s.write(char(v and 0xFF))

proc readExact(s: Stream; n: int): string =
  ## Read exactly n bytes from a pipe stream; raise on short read (EOF).
  result = newString(n)
  var got = 0
  while got < n:
    let r = s.readData(addr result[got], n - got)
    if r <= 0:
      raise newException(DecompressionError,
        "lzxd_helper daemon closed pipe after " & $got & "/" & $n & " bytes")
    got += r

proc readU32BE(s: Stream): uint32 =
  let h = readExact(s, 4)
  (uint32(h[0].byte) shl 24) or (uint32(h[1].byte) shl 16) or
    (uint32(h[2].byte) shl 8) or uint32(h[3].byte)

proc startDaemon*(): LzxDaemon =
  let helper = findHelper()
  let p = startProcess(helper, args = ["daemon"], options = {})
  result = LzxDaemon(process: p, inp: p.inputStream, outp: p.outputStream)

proc close*(d: var LzxDaemon) =
  if d.process != nil:
    try:
      if d.inp != nil: d.inp.close()   # EOF on stdin → daemon exits its loop
      discard d.process.waitForExit()
    except CatchableError: discard
    finally:
      d.process.close()
      d.process = nil

proc decodeLzx*(d: var LzxDaemon; stream: string; outLen: int): string =
  ## Decompress one stripped LZX bitstream via the helper daemon.
  d.inp.writeU32BE(uint32(outLen))
  d.inp.writeU32BE(uint32(stream.len))
  if stream.len > 0:
    d.inp.writeData(unsafeAddr stream[0], stream.len)
  d.inp.flush()

  let status = readU32BE(d.outp)
  let payloadLen = int(readU32BE(d.outp))
  let payload = readExact(d.outp, payloadLen)
  if status != 0:
    raise newException(DecompressionError, "lzxd_helper: " & payload.strip())
  if payload.len != outLen:
    raise newException(DecompressionError,
      "lzxd output length " & $payload.len & " != expected " & $outLen)
  return payload
